Adding documentation for schedules APIs (#4207)
pulumi/pulumi-hugo: themes/default/content/docs/pulumi-cloud/cloud-rest-api/_index.md

    "AWS_SESSION_TOKEN": "<redacted>""
```

## Schedules

> {{< notes >}}
> Pulumi Schedules REST API endpoints are currently in private preview.
> {{< /notes >}}

Reviewer: When we publish these docs that won't be true anymore.

```suggestion
```

| `stack`             | string | path  | stack name                                                                                                   |

#### Example

```bash
curl \
  -H "Accept: application/vnd.pulumi+8" \
  -H "Content-Type: application/json" \
  -H "Authorization: token $PULUMI_ACCESS_TOKEN" \
  https://api.pulumi.com/api/stacks/{organization}/{project}/{stack}/deployments/schedules
```

#### Default response

```
Status: 200 OK
```

```
{
  "scheduledDeployments": [
    {
      "id": "12345678-8102-447f-b246-e9ec85786e23",
      "orgID": "87654321-8b3b-418a-8c01-5ddd0e00bace",
      "scheduleCron": "0 0 * * *",
      "nextExecution": "2024-04-20 00:00:00.000",
      "paused": false,
      "kind": "deployment",
      "definition": {
        "programID": "12345678-9f94-4a92-8bfb-c2da724fa65b",
        "request": {
          "operation": "update"
        }
      },
      "created": "2024-04-19 01:00:00.000",
      "modified": "2024-04-19 01:00:00.000",
      "lastExecuted": null
    }
  ]
}

```

### Create Drift Schedule

> 

Reviewer: Would be great to add a note here and Update Drift Schedule and Create/Update TTL Schedule that says something like:

```
This is a convenience API for ease of use to create/update [this] schedule, but the same configuration is possible via the raw deployments schedule APIs. 
```

```
POST /api/stacks/{organization}/{project}/{stack}/deployments/drift/schedules
```

#### Parameters

| Parameter           | Type   | In    | Description                                                                                                  |
|---------------------|--------|-------|--------------------------------------------------------------------------------------------------------------|
| `organization`      | string | path  | organization name                                                                                            |
| `project`           | string | path  | project name                                                                                                 |
| `stack`             | string | path  | stack name                                                                                                   |
| `scheduleCron`      | string | body  | cron expression for when to run drift detection                                                              |

> | `autoRemediate`     | bool   | body  | whether to remediate any detected drift                                                                      |

Reviewer 2: ```suggestion
| `autoRemediate`     | bool   | body  | true if detected drift should be remediated automatically                                                                     |
```

        "operation": "detect-drift",
        "operationContext": {
            "options": {
                "remediateIfDriftDetected": true
            }
        }
    }
  },
  "created": "2024-04-19 01:00:00.000",
  "modified": "2024-04-19 01:00:00.000",
  "lastExecuted": null
}
```

### Create TTL Schedule

```
POST /api/stacks/{organization}/{project}/{stack}/deployments/ttl/schedules
```

#### Parameters

| Parameter           | Type   | In    | Description                                                                                                  |
|---------------------|--------|-------|--------------------------------------------------------------------------------------------------------------|
| `organization`      | string | path  | organization name                                                                                            |
| `project`           | string | path  | project name                                                                                                 |
| `stack`             | string | path  | stack name                                                                                                   |

> | `timestamp`         | string | body  | ISO timestamp of when to destroy the stack                                                                   |

Reviewer: Can we put an example timestamp here

> | `timestamp`         | string | body  | ISO timestamp of when to destroy the stack                                                                   |

Reviewer 2: ```suggestion
| `timestamp`         | string | body  | ISO 8601 timestamp specifying when to destroy the stack                                                                   |
```

> | `deleteAfterDestroy`| bool   | body  | whether to delete the stack after resources are destroyed                                                    |

Reviewer 2: ```suggestion
| `deleteAfterDestroy`| bool   | body  | true if the stack should be deleted after resources are destroyed                                                    |
```

#### Example

```bash
curl \
  -H "Accept: application/vnd.pulumi+8" \
  -H "Content-Type: application/json" \
  -H "Authorization: token $PULUMI_ACCESS_TOKEN" \
  --request POST \

>   --data '{"timestamp":"{timestampValue}","deleteAfterDestroy":"{deleteAfterDestroyValue}"}' \

Reviewer: Make it a real example, so fill in the values instead of `{timestampValue}` and `{deleteAfterDestroyValue}`. Applies throughout.

Author: Throughout the doc, I saw examples of both and wasn't sure which one was best. Updating to real values.

  https://api.pulumi.com/api/stacks/{organization}/{project}/{stack}/deployments/ttl/schedules
```

#### Default response

```
Status: 200 OK
```

```
{
  "id": "12345678-8102-447f-b246-e9ec85786e23",
  "orgID": "87654321-8b3b-418a-8c01-5ddd0e00bace",

>   "scheduleCron": "0 0 * * *",

Reviewer 2: This looks wrong to me. I think `scheduleOnce` should be set instead.

Author: That is correct, great catch! Fixed here and in Update TTL

  "nextExecution": "2024-04-20 00:00:00.000",
  "paused": false,
  "kind": "deployment",
  "definition": {
    "programID": "12345678-9f94-4a92-8bfb-c2da724fa65b",
    "request": {
        "inheritSettings": true,
        "operation": "destroy",
        "operationContext": {
            "options": {
                "deleteAfterDestroy": true
            }
        }
    }
  },
  "created": "2024-04-19 01:00:00.000",
  "modified": "2024-04-19 01:00:00.000",
  "lastExecuted": null
}
```

> ### Create Raw Operation Schedule

Reviewer: ```suggestion
### Create Raw Deployment Schedule
```

```
POST /api/stacks/{organization}/{project}/{stack}/deployments/schedules
```

#### Parameters

| Parameter           | Type   | In    | Description                                                                                                  |
|---------------------|--------|-------|--------------------------------------------------------------------------------------------------------------|
| `organization`      | string | path  | organization name                                                                                            |
| `project`           | string | path  | project name                                                                                                 |
| `stack`             | string | path  | stack name                                                                                                   |
| `scheduleCron`      | string | body  | cron expression for when to run the pulumi operation                                                         |

> | `scheduleOnce`      | string | body  | ISO timestamp of when to destroy the stack                                                                   |

Reviewer 2: ```suggestion
| `scheduleOnce`      | string | body  | ISO 8601 timestamp specifying when to destroy the stack                                                                   |
```

> | `operation`         | string | body  | pulumi operation to perform. Can be "update", "destroy", "refresh" or "preview"                              |

Reviewer: This should be `request` instead of `operation` I think?
```suggestion
| `request`         | [CreateDeploymentRequest](/docs/pulumi-cloud/deployments/api/#create-deployment) | body  | The create deployment request object that will be executed on every invocation |
```

Author: I was aiming to make it simpler for users, and to match UI, but yeah, ideally we should point them to full range of things for the raw one, updating to this.

> Note: only pass in either `scheduleCron` or `scheduleOnce`, and not both. One of them is required though.

Reviewer: ```suggestion
Note: Only one of `scheduleCron` and `scheduleOnce` must be set.
```

> Note: only pass in either `scheduleCron` or `scheduleOnce`, and not both. One of them is required though.

Reviewer 2: ```suggestion
Note: Exactly one of `scheduleCron` and `scheduleOnce` must be set.
```

#### Example

```bash
curl \
  -H "Accept: application/vnd.pulumi+8" \
  -H "Content-Type: application/json" \
  -H "Authorization: token $PULUMI_ACCESS_TOKEN" \
  --request POST \
  --data '{ "scheduleCron":"{scheduleCronValue}", "request": { "operation": "{operationValue}" } }' \
  https://api.pulumi.com/api/stacks/{organization}/{project}/{stack}/deployments/schedules
```

> The `request` field is really a [CreateDeploymentRequest](/docs/pulumi-cloud/deployments/api/#create-deployment) under the hood. You can pass in the same parameters to customize the scheduled deployment, but they are not required.

Reviewer: Remove this section since we've explained the usage in the parameters section.
```suggestion
```

> The `request` field is really a [CreateDeploymentRequest](/docs/pulumi-cloud/deployments/api/#create-deployment) under the hood. You can pass in the same parameters to customize the scheduled deployment, but they are not required.

Reviewer 2: ```suggestion
The `request` field is a [CreateDeploymentRequest](/docs/pulumi-cloud/deployments/api/#create-deployment), which optionally allows full customization of the deployment to be scheduled.
```

}
```

### Update Drift Schedule

```
POST /api/stacks/{organization}/{project}/{stack}/deployments/drift/schedules/{scheduleID}
```

| Parameter           | Type   | In    | Description                                                                                                  |
|---------------------|--------|-------|--------------------------------------------------------------------------------------------------------------|
| `organization`      | string | path  | organization name                                                                                            |
| `project`           | string | path  | project name                                                                                                 |
| `stack`             | string | path  | stack name                                                                                                   |
| `scheduleID`        | string | path  | schedule ID that you want to update                                                                          |
| `scheduleCron`      | string | body  | cron expression for when to run drift detection                                                              |

> | `autoRemediate`     | bool   | body  | whether to remediate any detected drift                                                                      |

Reviewer 2: ```suggestion
| `autoRemediate`     | bool   | body  | true if detected drift should be remediated automatically                                                                      |
```

}
```

### Update TTL Schedule

```
POST /api/stacks/{organization}/{project}/{stack}/deployments/ttl/schedules/{scheduleID}
```

| Parameter           | Type   | In    | Description                                                                                                  |
|---------------------|--------|-------|--------------------------------------------------------------------------------------------------------------|
| `organization`      | string | path  | organization name                                                                                            |
| `project`           | string | path  | project name                                                                                                 |
| `stack`             | string | path  | stack name                                                                                                   |
| `scheduleID`        | string | path  | schedule ID that you want to update                                                                          |
| `timestamp`         | string | body  | ISO timestamp of when to destroy the stack                                                                   |

> | `deleteAfterDestroy`| bool   | body  | whether to delete the stack after resources are destroyed                                                    |

Reviewer 2: ```suggestion
| `deleteAfterDestroy`| bool   | body  | true if the stack should be deleted after resources are destroyed                                                    |
```

#### Example

```bash
curl \
  -H "Accept: application/vnd.pulumi+8" \
  -H "Content-Type: application/json" \
  -H "Authorization: token $PULUMI_ACCESS_TOKEN" \
  --request POST \
  --data '{"timestamp":"{timestampValue}","deleteAfterDestroy":"{deleteAfterDestroyValue}"}' \
  https://api.pulumi.com/api/stacks/{organization}/{project}/{stack}/deployments/ttl/schedules/{scheduleID}
```

#### Default response

```
Status: 200 OK
```

```
{
  "id": "12345678-8102-447f-b246-e9ec85786e23",
  "orgID": "87654321-8b3b-418a-8c01-5ddd0e00bace",
  "scheduleCron": "0 0 * * *",
  "nextExecution": "2024-04-20 00:00:00.000",
  "paused": false,
  "kind": "deployment",
  "definition": {
    "programID": "12345678-9f94-4a92-8bfb-c2da724fa65b",
    "request": {
        "inheritSettings": true,
        "operation": "destroy",
        "operationContext": {
            "options": {
                "deleteAfterDestroy": true
            }
        }
    }
  },
  "created": "2024-04-19 01:00:00.000",
  "modified": "2024-04-19 01:00:00.000",
  "lastExecuted": null
}
```

> ### Update Raw Operation Schedule

Reviewer: ```suggestion
### Update Raw Deployment Schedule
```

```
POST /api/stacks/{organization}/{project}/{stack}/deployments/schedules/{scheduleID}
```

#### Parameters

| Parameter           | Type   | In    | Description                                                                                                  |
|---------------------|--------|-------|--------------------------------------------------------------------------------------------------------------|
| `organization`      | string | path  | organization name                                                                                            |
| `project`           | string | path  | project name                                                                                                 |
| `stack`             | string | path  | stack name                                                                                                   |
| `scheduleID`        | string | path  | schedule ID that you want to update                                                                          |
| `scheduleCron`      | string | body  | cron expression for when to run the pulumi operation                                                         |

> | `scheduleOnce`      | string | body  | ISO timestamp of when to destroy the stack                                                                   |

Reviewer 2: ```suggestion
| `scheduleOnce`      | string | body  | ISO 8601 timestamp specifying when to destroy the stack                                                                   |
```

| `operation`         | string | body  | pulumi operation to perform. Can be "update", "destroy", "refresh" or "preview"                              |

> Note: only pass in either `scheduleCron` or `scheduleOnce`, and not both. One of them is required though.

Reviewer 2: ```suggestion
Note: Exactly one of `scheduleCron` and `scheduleOnce` must be set.
```

#### Example

```bash
curl \
  -H "Accept: application/vnd.pulumi+8" \
  -H "Content-Type: application/json" \
  -H "Authorization: token $PULUMI_ACCESS_TOKEN" \
  --request POST \
  --data '{ "scheduleCron":"{scheduleCronValue}", "request": { "operation": "{operationValue}" } }' \
  https://api.pulumi.com/api/stacks/{organization}/{project}/{stack}/deployments/schedules/{scheduleID}
```

> The `request` field is really a [CreateDeploymentRequest](/docs/pulumi-cloud/deployments/api/#create-deployment) under the hood. You can pass in the same parameters to customize the scheduled deployment, but they are not required.

Reviewer: Same feedback as in Create Raw Deployment Schedule

> The `request` field is really a [CreateDeploymentRequest](/docs/pulumi-cloud/deployments/api/#create-deployment) under the hood. You can pass in the same parameters to customize the scheduled deployment, but they are not required.

Reviewer 2: ```suggestion
The `request` field is a [CreateDeploymentRequest](/docs/pulumi-cloud/deployments/api/#create-deployment), which optionally allows full customization of the deployment to be scheduled.
```

  "paused": false,
  "kind": "deployment",
  "definition": {
      "programID": "12345678-911a-43a3-8389-109be345b1d6",
      "request": {
          "operation": "update"
      }
  },
  "created": "2024-04-19 01:00:00.000",
  "modified": "2024-04-19 01:00:00.000",
  "lastExecuted": null
}
```

### Delete Schedule

```
DELETE /api/stacks/{organization}/{project}/{stack}/deployments/schedules/{scheduleID}
```

| Parameter           | Type   | In    | Description                                                                                                  |
|---------------------|--------|-------|--------------------------------------------------------------------------------------------------------------|
| `organization`      | string | path  | organization name                                                                                            |
| `project`           | string | path  | project name                                                                                                 |
| `stack`             | string | path  | stack name                                                                                                   |

> | `scheduleID`        | string | path  | schedule ID that you want to get                                                                             |

Reviewer 2: ```suggestion
| `scheduleID`        | string | path  | schedule ID that you want to delete                                                                             |
```

    "AWS_SESSION_TOKEN": "<redacted>""
```

## Schedules

> {{< notes >}}
> Pulumi Schedules REST API endpoints are currently in public preview.
> {{< /notes >}}

Reviewer: Remove this please!
```suggestion
```
